pkg/operator/controller/instance_controller.go

```diff
@@ -0,0 +1,742 @@
+func (r *InstanceReconciler) setPVC(ctx context.Context, name string) error {
+	// Create the PVC for cl-controlplane
+	controlplanePVC := &corev1.PersistentVolumeClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: r.clNamespace,
+		},
+		Spec: corev1.PersistentVolumeClaimSpec{
+			AccessModes: []corev1.PersistentVolumeAccessMode{corev1.ReadWriteOnce},
+			Resources: corev1.VolumeResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceStorage: resource.MustParse("100Mi"),
+				},
+			},
+		},
+	}
+
+	return r.createResource(ctx, controlplanePVC)
+}
+
+// setRules sets up k8s ClusterRule and ClusterRoleBinding for the controlplane.
+func (r *InstanceReconciler) setRules(ctx context.Context, name string) error {
+	// Create the ClusterRole for the controlplane.
+	clusterRole := &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: r.clNamespace,
+		},
+		Rules: []rbacv1.PolicyRule{
+			{
+				APIGroups: []string{""},
+				Resources: []string{"services"},
+				Verbs:     []string{"create", "delete", "update"},
+			},
+			{
+				APIGroups: []string{""},
+				Resources: []string{"pods"},
+				Verbs:     []string{"get", "list", "watch"},
+			},
+		},
+	}
+
+	err := r.createResource(ctx, clusterRole)
+	if err != nil {
+		return err
+	}
+
+	// Create ClusterRoleBinding for the controlplane.
+	clusterRoleBinding := &rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: r.clNamespace,
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "ClusterRole",
+			Name:     name,
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      "default",
+				Namespace: "default",
+			},
+		},
+	}
+	return r.createResource(ctx, clusterRoleBinding)
+}
+
+// setNameSpace sets up k8s ClusterRule and ClusterRoleBinding for the controlplane.
+func (r *InstanceReconciler) setNameSpace(ctx context.Context) error {
+	ns := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: r.clNamespace,
+		},
+	}
+
+	return r.createResource(ctx, ns)
+}
+
+// setNameSpace sets up the external service for the project.
+func (r *InstanceReconciler) setExternalService(ctx context.Context) error {
+	// Create a Service object
+	service := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      IngressName,
+			Namespace: r.clNamespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: map[string]string{
+				"app": "cl-dataplane",
+			},
+			Ports: []corev1.ServicePort{
+				{
+					Port:       dpapi.ListenPort,
+					TargetPort: intstr.FromInt(dpapi.ListenPort),
+					Protocol:   corev1.ProtocolTCP,
+					Name:       "http",
+				},
+			},
+		},
+	}
+
+	switch r.instance.Spec.Ingress.Type {
+	case api.IngressTypeNone:
+		return nil
+	case api.IngressTypeNodePort:
+		service.Spec.Type = corev1.ServiceTypeNodePort
+		service.Spec.Ports[0].NodePort = ExternalNodePort
```

Comment:
do we want to hard code the port or let k8s report it for us (and we can reflect in the status)?
If the port is taken - this will fail - in which case you need to reflect back to the Instance.Status.